File: Solution/Programmers/순위 검색.py
from itertools import combinations
from bisect import bisect_left

def solution(info, query):
    info_dict = {}
    for i in info:
        information = i.split(' ')
        keys = information[:-1]
        value = int(information[-1])
        for n in range(5):
            key_comb = list(combinations(keys, n))
            for k in key_comb:
                key = ''.join(k)
                if info_dict.get(key) is None:
                    info_dict[key] = [value]
                else:
                    info_dict[key].append(value)
                    
    for key in info_dict.keys():
        info_dict[key] = sorted(info_dict[key])
    
    result = []
    for q in query:
        q = q.replace('and ', '')
        q = q.split(' ')
        q_keys = q[:-1]
        q_value = int(q[-1])
        
        while '-' in q_keys:
            q_keys.remove('-')
        
        q_key = ''.join(q_keys)
        if not q_key in info_dict:
            result.append(0)
        else:
            values = info_dict[q_key]
            #print(values)
            result.append(len(values) - bisect_left(values, q_value))
    
    return result

File: Solution/Programmers/순위 검색.js
const getCombinations = function (arr, selectNumber) {
  const results = [];
  if (selectNumber === 1) return arr.map(value => [value]);

  arr.forEach((fixed, index, origin) => {
    const rest = origin.slice(index + 1);
    const combinations = getCombinations(rest, selectNumber - 1);
    const attached = combinations.map(combination => [fixed, ...combination]);
    results.push(...attached);
  });

  return results;
};

function solution(infos, queries) {
  let dict = {};
  let result = [];
  for (let info of infos) {
    info = info.split(" ");
    const keys = info.slice(0, 4);
    const score = info.slice(4, 5);
    for (let i = 0; i < 5; i++) {
      const keyList = getCombinations(keys, i);
      if (keyList.length === 0) {
        keyList.push(" ");
        const key = " ";
        if (key in dict) {
          dict[key].push(Number(score));
        } else {
          dict[key] = [Number(score)];
        }
      } else {
        for (let k of keyList) {
          const key = k.join("");
          if (key in dict) {
            dict[key].push(Number(score));
          } else {
            dict[key] = [Number(score)];
          }
        }
      }
    }
  }

  for (let [key, value] of Object.entries(dict)) {
    const sortedValue = value.sort((a, b) => a - b);
    dict[key] = sortedValue;
  }

  for (let query of queries) {
    query = query.replace(/ and/g, "").replace(/\-/g, "").split(" ");
    let key = query.slice(0, 4);
    const score = Number(query.slice(4, 5));

    key = key.join("") === "" ? " " : key.join("");
    if (key in dict) {
      const values = dict[key];
      let left = 0;
      let right = values.length;
      let temp = 0;
      while (left < right) {
        const mid = Math.floor((left + right) / 2);
        if (values[mid] >= score) {
          right = mid;
          temp = mid;
        } else {
          left = mid + 1;
        }
      }
      result.push(values.length - temp);
    } else {
      result.push(0);
    }
  }
  return result;
}
